pkg/controller/snapshotter/snapshotter.go

```diff
@@ -160,78 +166,64 @@ func (vs *volumeSnapshotter) getSnapshotDataFromSnapshotName(snapshotName string
-	var snapshotDataObj crdv1.VolumeSnapshotData
-	snapshotDataName := snapshot.Spec.SnapshotDataName
-	glog.Infof("In waitForPluginSnapshot: %s", snapshotName)
+func (vs *volumeSnapshotter) waitForSnapshot(snapshotName string, snapshot *crdv1.VolumeSnapshot, snapshotDataObj *crdv1.VolumeSnapshotData) error {
+	// Get a fresh VolumeSnapshot from API
+	var snapshotObj crdv1.VolumeSnapshot
 	err := vs.restClient.Get().
-		Name(snapshotDataName).
-		Resource(crdv1.VolumeSnapshotDataResourcePlural).
-		Namespace(v1.NamespaceDefault).
-		Do().Into(&snapshotDataObj)
-	if err != nil {
-		return err
-	}
-
-	pv, err := vs.getPVFromVolumeSnapshot(snapshotName, snapshot)
+		Name(snapshot.Metadata.Name).
+		Resource(crdv1.VolumeSnapshotResourcePlural).
+		Namespace(snapshot.Metadata.Namespace).
+		Do().Into(&snapshotObj)
 	if err != nil {
-		return err
+		return fmt.Errorf("Error getting snapshot object %s from API server: %v", snapshotName, err)
 	}
-	spec := &pv.Spec
-	volumeType := crdv1.GetSupportedVolumeFromPVSpec(spec)
-	if len(volumeType) == 0 {
-		return fmt.Errorf("unsupported volume type found in PV %#v", spec)
+	snapshotDataName := snapshotObj.Spec.SnapshotDataName
+	glog.Infof("In waitForSnapshot: snapshot %s snapshot data %s", snapshotName, snapshotDataName)
+	if snapshotDataObj == nil {
+		return fmt.Errorf("Failed to update VolumeSnapshot for snapshot %s: no VolumeSnapshotData.", snapshotName)
 	}
-	plugin, ok := (*vs.volumePlugins)[volumeType]
-	if !ok {
-		return fmt.Errorf("%s is not supported volume for %#v", volumeType, spec)
+
+	plugin, err := vs.getPlugin(snapshotName, snapshot)
+	if plugin == nil {
+		return fmt.Errorf("Failed to get volume plugin to wait for snapshot creation %s. %v", snapshotName, err)
 	}
 
-	var completed bool = false
+	var newSnapshot *crdv1.VolumeSnapshot = nil
 	var conditions *[]crdv1.VolumeSnapshotCondition = nil
+	var status, newstatus string = "", ""
 	// Wait until the snapshot is successfully created by the plugin or an error occurs that
 	// fails the snapshot creation.
 	for {
-		// TODO(xyang): Use status returned from DescribeSnapshot to update snapshot
-		conditions, completed, err = plugin.DescribeSnapshot(&snapshotDataObj)
-		if !completed {
-			glog.Infof("Snapshot %s creation is not complete yet. Status: [%#v] Retrying...", snapshotName, conditions)
-			if err != nil {
-				// Don't bail out here. The error message could be "current snapshot status is: creating".
-				glog.Infof("Message received when creating snapshot %s: %v", snapshotName, err)
+		conditions, _, err = (*plugin).DescribeSnapshot(snapshotDataObj)
+		newstatus = vs.getSimplifiedSnapshotStatus(conditions)
+		if newstatus == statusPending {
+			if newstatus != status {
+				status = newstatus
+				glog.Infof("Snapshot %s creation is not complete yet. Status: [%#v] Retrying...", snapshotName, conditions)
+				// UpdateVolmeSnapshot status
+				newSnapshot, err = vs.UpdateVolumeSnapshot(snapshotName, conditions)
```

Comment:
The condition is an array of string? But I see DescribeSnapshot() only returns a single string for status?

Reply:
Looks like rootfs has answered this question.  Will append the new condition to existing conditions.

Reply:
Modified UpdateVolumeSnapshot and UpdateVolumeSnapshotData to append the new condition to existing ones.

Currently DescribeSnapshot returns *[]crdv1.VolumeSnapshotCondition. Should this be changed to *crdv1.VolumeSnapshotCondition instead because DescribeSnapshot should just have one status, not multiples.